File: coletores/bnp_scraper.py
# coletores/bnp_scraper.py - Módulo para o scraper do Pangea BNP com Selenium em Python

import time
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.by import By
import traceback

def scrape_bnp(termo_de_busca):
    """
    Controla um navegador headless para buscar e extrair precedentes do Pangea BNP.
    """
    print("\n--- INICIANDO SCRAPER PANGEA BNP (Python/Selenium) ---")
    
    chrome_options = Options()
    chrome_options.add_argument("--headless=new")
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
    chrome_options.add_argument("--disable-gpu")
    chrome_options.add_argument("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/5.37.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36")

    driver = None
    try:
        # O chromedriver está no PATH do sistema, o Selenium o encontrará.
        driver = webdriver.Chrome(options=chrome_options)
        documentos = []
        url = f"https://pangeabnp.pdpj.jus.br/precedentes?q={termo_de_busca}"
        print(f"Acessando: {url}")
        driver.get(url)

        print("Aguardando o carregamento dinâmico dos resultados...")
        # Espera explícita para garantir que os resultados carreguem
        WebDriverWait(driver, 30).until(
            EC.presence_of_element_located((By.TAG_NAME, "app-card-precedente-item"))
        )
        print("Página carregada, iniciando extração.")

        page_source = driver.page_source
        soup = BeautifulSoup(page_source, 'html.parser')

        resultados = soup.find_all('app-card-precedente-item')
        print(f"Encontrados {len(resultados)} resultados na página.")

        for item in resultados:
            try:
                titulo_tag = item.find('h5', class_='card-title')
                link_tag = item.find('a', class_='card-title-link')
                if not (titulo_tag and link_tag): continue

                titulo = titulo_tag.get_text(strip=True)
                link = link_tag['href']

                detalhes = {
                    dl.find('dt').get_text(strip=True): dl.find('dd').get_text(strip=True)
                    for dl in item.find_all('dl') if dl.find('dt') and dl.find('dd')
                }
                
                ementa_tag = item.find('p', class_='card-text')
                ementa = ementa_tag.get_text(strip=True) if ementa_tag else ''
                original_date_str = detalhes.get('Data de Julgamento:', '')
                numero_unico = detalhes.get('Número Único:', '')
                doc_id = f"BNP-{numero_unico}" if numero_unico else f"BNP-{hash(titulo)}"

                documento = {
                    "tipo_documento": "precedente", "id": doc_id, "titulo": titulo,
                    "link": f"https://pangeabnp.pdpj.jus.br{link}", "fonte": "Pangea BNP",
                    "data_julgamento": original_date_str,
                    "orgaoJulgador": detalhes.get('Órgão Julgador:', ''),
                    "ramoDireito": detalhes.get('Ramo do Direito:', ''),
                    "numeroUnico": numero_unico, "assuntos": detalhes.get('Assuntos:', ''), "ementa": ementa
                }
                documentos.append(documento)
            except Exception as e:
                print(f"Erro ao processar um item individual do BNP: {e}")
        
        print(f"--- SCRAPER BNP FINALIZADO --- Total coletado: {len(documentos)}")
        return documentos
    
    except Exception as e:
        print(f"ERRO CRÍTICO no scraper do BNP: {e}")
        traceback.print_exc()
        return []
    finally:
        if driver:
            driver.quit()
            print("Navegador Selenium finalizado.")

File: coletores/bnp_scraper.js
// coletores/bnp_scraper.js - Coletor para Pangea BNP em Node.js com Selenium

const { Builder, By, until } = require('selenium-webdriver');
const chrome = require('selenium-webdriver/chrome');
const cheerio = require('cheerio');

async function scrapeBnp(termoDeBusca) {
    console.error("\n--- INICIANDO SCRAPER PANGEA BNP (JS) ---");
    
    const options = new chrome.Options();
    options.addArguments('--headless=new', '--no-sandbox', '--disable-dev-shm-usage', '--disable-gpu');
    options.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36');

    let driver;
    try {
        driver = await new Builder()
            .forBrowser('chrome')
            .setChromeOptions(options)
            .build();

        const url = `https://pangeabnp.pdpj.jus.br/precedentes?q=${encodeURIComponent(termoDeBusca)}`;
        console.error(`Acessando: ${url}`);
        await driver.get(url);

        console.error("Aguardando o carregamento dinâmico dos resultados...");
        await driver.wait(until.elementLocated(By.tagName("app-card-precedente-item")), 30000);
        console.error("Página carregada, iniciando extração.");
        
        const pageSource = await driver.getPageSource();
        const $ = cheerio.load(pageSource);

        const resultados = $('app-card-precedente-item');
        console.error(`Encontrados ${resultados.length} resultados na página.`);
        const documentos = [];

        resultados.each((i, item) => {
            const tituloTag = $(item).find('h5.card-title');
            const linkTag = $(item).find('a.card-title-link');
            if (!tituloTag.length || !linkTag.length) return;

            const detalhes = {};
            $(item).find('dl').each((idx, dl) => {
                const dt = $(dl).find('dt').text().trim();
                const dd = $(dl).find('dd').text().trim();
                if (dt && dd) detalhes[dt] = dd;
            });
            
            const ementa = $(item).find('p.card-text').text().trim() || '';
            const numeroUnico = detalhes['Número Único:'] || '';
            const titulo = tituloTag.text().trim();

            const documento = {
                tipo_documento: "precedente",
                id: numeroUnico ? `BNP-${numeroUnico}` : `BNP-${Buffer.from(titulo).toString('hex')}`,
                titulo: titulo,
                link: `https://pangeabnp.pdpj.jus.br${linkTag.attr('href')}`,
                fonte: "Pangea BNP",
                data_julgamento: detalhes['Data de Julgamento:'] || '',
                orgaoJulgador: detalhes['Órgão Julgador:'] || '',
                ramoDireito: detalhes['Ramo do Direito:'] || '',
                numeroUnico: numeroUnico,
                assuntos: detalhes['Assuntos:'] || '',
                ementa: ementa
            };
            documentos.push(documento);
        });

        console.error(`--- SCRAPER BNP (JS) FINALIZADO --- Total coletado: ${documentos.length}`);
        console.log(JSON.stringify(documentos));

    } catch (error) {
        console.error(`ERRO CRÍTICO no scraper do BNP (JS): ${error.message}`);
        console.log(JSON.stringify([]));
    } finally {
        if (driver) {
            await driver.quit();
            console.error("Navegador Selenium finalizado.");
        }
    }
}

if (require.main === module) {
    const termo = process.argv[2];
    if (!termo) {
        console.error("Erro: Nenhum termo de busca fornecido.");
        process.exit(1);
    }
    scrapeBnp(termo);
}